flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java

```diff
@@ -0,0 +1,356 @@
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * The general buffer manager used by {@link InputChannel} to request/recycle
+ * exclusive or floating buffers.
+ */
+public class BufferManager implements BufferListener, BufferRecycler {
+
+	/** The available buffer queue wraps both exclusive and requested floating buffers. */
+	private final AvailableBufferQueue bufferQueue = new AvailableBufferQueue();
+
+	/** The buffer provider for requesting exclusive buffers. */
+	private final MemorySegmentProvider globalPool;
+
+	/** The input channel to own this buffer manager. */
+	private final InputChannel inputChannel;
+
+	/** The tag indicates whether it is waiting for additional floating buffers from the buffer pool. */
+	@GuardedBy("bufferQueue")
+	private boolean isWaitingForFloatingBuffers;
+
+	/** The total number of required buffers for the respective input channel. */
+	@GuardedBy("bufferQueue")
+	private int numRequiredBuffers;
+
+	public BufferManager(
+		MemorySegmentProvider globalPool,
+		InputChannel inputChannel,
+		int numRequiredBuffers) {
+
+		this.globalPool = checkNotNull(globalPool);
+		this.inputChannel = checkNotNull(inputChannel);
+		checkArgument(numRequiredBuffers >= 0);
+		this.numRequiredBuffers = numRequiredBuffers;
+	}
+
+	// ------------------------------------------------------------------------
+	// Buffer request
+	// ------------------------------------------------------------------------
+
+	@Nullable
+	Buffer requestBuffer() {
+		synchronized (bufferQueue) {
+			return bufferQueue.takeBuffer();
+		}
+	}
+
+	/**
+	 * Requests exclusive buffers from the provider and returns the number of requested amount.
+	 */
+	int requestExclusiveBuffers() throws IOException {
+		Collection<MemorySegment> segments = globalPool.requestMemorySegments();
+		checkArgument(!segments.isEmpty(), "The number of exclusive buffers per channel should be larger than 0.");
+
+		synchronized (bufferQueue) {
+			for (MemorySegment segment : segments) {
+				bufferQueue.addExclusiveBuffer(new NetworkBuffer(segment, this), numRequiredBuffers);
+			}
+		}
+		return segments.size();
+	}
+
+	/**
+	 * Requests floating buffers from the buffer pool based on the given required amount, and returns the actual
+	 * requested amount. If the required amount is not fully satisfied, it will register as a listener.
+	 */
+	int requestFloatingBuffers(int numRequired) throws IOException {
+		int numRequestedBuffers = 0;
+		synchronized (bufferQueue) {
+			// Similar to notifyBufferAvailable(), make sure that we never add a buffer after channel
+			// released all buffers via releaseAllResources().
+			if (inputChannel.isReleased()) {
+				return numRequestedBuffers;
+			}
+
+			numRequiredBuffers = numRequired;
+
+			while (bufferQueue.getAvailableBufferSize() < numRequiredBuffers && !isWaitingForFloatingBuffers) {
+				BufferPool bufferPool = inputChannel.inputGate.getBufferPool();
```

Comment:
I also hope so, but the truth is not, because we can not get the proper `BufferPool` while constructing the `BufferManager`. 

The `BufferPool` is created in `SingleInputGate#setup`, and it is also the main reason of relying on `InputChannel` component in `BufferManager` to bypass this issue.